CMakeLists.txt

```diff
@@ -0,0 +1,91 @@
+    src/error_handler.c
+)
+
+set(HEADERS
+    src/stm32f4xx_hal_conf.h
+    src/defines.h
+    src/enc28j60.h
+    src/ip_arp_udp_tcp.h
+    src/net.h
+    src/dhcp.h
+    src/dnslkup.h
+    src/websrv_help_functions.h
+    src/EtherShield.h
+    src/error_handler.h
+)
+
+set(ETHERNET_LED_GPIO       "GPIOA"         CACHE INTERNAL "GPIO for ethernet LED")
+set(ETHERNET_LED_PIN        "GPIO_PIN_2"    CACHE INTERNAL "PIN for ethernet LED")
+set(ETHERNET_CS_GPIO        "GPIOA"         CACHE INTERNAL "GPIO for SPI chip select")
+set(ETHERNET_CS_PIN         "GPIO_PIN_3"    CACHE INTERNAL "PIN for SPI chip select")
+set(ETHERNET_CS_DELAY       "2"             CACHE INTERNAL "chip-select delay for ehternet")
+
+set(UDP_client              "1"                     CACHE INTERNAL "enables UDP transport protocol")
+set(DNS_client              "1"                     CACHE INTERNAL "enables DNS client")
+set(DHCP_client             "1"                     CACHE INTERNAL "enables DHCP client")
+set(PING_client             "1"                     CACHE INTERNAL "enables ICMP ping client")
+set(PINGPATTERN             0x42                    CACHE INTERNAL "ping pattern value")
+set(TCP_client              "1"                     CACHE INTERNAL "enables TCP transport protocol")
+set(WWW_client              "1"                     CACHE INTERNAL "enables HTTP")
+set(WWW_USER_AGENT          "stm32-enc28j60/1.0"    CACHE INTERNAL "HTTP User-Agent")
+set(FROMDECODE_websrv_help  "1"                     CACHE INTERNAL "")
+set(URLENCODE_websrv_help   "1"                     CACHE INTERNAL "")
+
+set(CMAKE_INCLUDE_CURRENT_DIR TRUE)
+
+list(APPEND HAL_COMP_LIST STM32F4 CORTEX RCC GPIO SPI)
+list(APPEND CMSIS_COMP_LIST "STM32F4")
+
+if(NOT STM32_CMAKE_PACKAGES)
+    find_package(CMSIS COMPONENTS "${CMSIS_COMP_LIST}" REQUIRED)
+    find_package(HAL COMPONENTS "${HAL_COMP_LIST}" REQUIRED)
+    set(STM32_CMAKE_PACKAGES 1)
+endif()
+
+add_library(stm32-enc28j60 ${SOURCES} ${HEADERS})
+target_include_directories(stm32-enc28j60 PUBLIC src)
+target_compile_definitions(stm32-enc28j60 PUBLIC
+    ETHERNET_LED_GPIO=${ETHERNET_LED_GPIO}
+    ETHERNET_LED_PIN=${ETHERNET_LED_PIN}
+    ETHERNET_CS_GPIO=${ETHERNET_CS_GPIO}
+    ETHERNET_CS_PIN=${ETHERNET_CS_PIN}
+    ETHERNET_CS_DELAY=${ETHERNET_CS_DELAY}
+
+    UDP_client=${UDP_client}
+    DNS_client=${DNS_client}
+    DHCP_client=${DHCP_client}
+    PING_client=${PING_client}
+    PINGPATTERN=${PINGPATTERN}
+    TCP_client=${TCP_client}
+    WWW_client=${WWW_client}
+    WWW_USER_AGENT="${WWW_USER_AGENT}"
+    FROMDECODE_websrv_help=${FROMDECODE_websrv_help}
+    URLENCODE_websrv_help=${URLENCODE_websrv_help}
+)
+
+target_link_libraries(stm32-enc28j60
+    HAL::STM32::F4::RCC
+    HAL::STM32::F4::GPIO
+    HAL::STM32::F4::CORTEX
+    CMSIS::STM32::F401VE
+    STM32::NoSys
+)
+
+#add_custom_command(TARGET stm32-enc28j60 POST_BUILD
+#    COMMAND ${CMAKE_OBJCOPY} -O ihex stm32-enc28j60.elf stm32-enc28j60.hex
+#)
```

Comment:
Out of curiosity, what is the purpose of these lines? :)

Reply:
its command for converting ELF to intel HEX, because some simulators like Proteus run only hex files. You need that command only on final firmware target